storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -2812,6 +2816,76 @@ fseg_free_extent(
 	return DB_SUCCESS;
 }
 
+/** Free the extent and fragment page associated with
+the segment.
+@param iblock     page where segment header are placed
+@param inode      index node information
+@param space      tablespace where segment resides
+@param mtr        mini-transaction
+@param header     segment header
+@param free_inode free the inode page also
+@param ahi        adaptive hash index
+@return DB_SUCCESS_LOCKED_REC when freeing wasn't completed
+@return DB_SUCCESS or other error code  when freeing was completed */
+static
+dberr_t fseg_free_step_low(buf_block_t *iblock, fseg_inode_t *inode,
+                           fil_space_t *space, mtr_t *mtr,
+                           fseg_header_t *header, bool free_inode
+#ifdef BTR_CUR_HASH_ADAPT
+                           , bool ahi=false
+#endif /* BTR_CUR_HASH_ADAPT */
+                           )
+{
+  dberr_t err= DB_SUCCESS;
+  if (xdes_t *descr= fseg_get_first_extent(inode, space, mtr, &err))
+  {
+    err= fseg_free_extent(inode, iblock, space,
+                          xdes_get_offset(descr), mtr
+#ifdef BTR_CUR_HASH_ADAPT
+                          , ahi
+#endif /* BTR_CUR_HASH_ADAPT */
+                        );
+    if (err == DB_SUCCESS)
+      err= DB_SUCCESS_LOCKED_REC;
+  }
+
+  if (err != DB_SUCCESS)
+    return err;
+
+  /* Free a frag page */
+  ulint n = fseg_find_last_used_frag_page_slot(inode);
+  if (UNIV_UNLIKELY(n == ULINT_UNDEFINED))
+  {
+    if (free_inode)
+      fsp_free_seg_inode(space, inode, iblock, mtr);
+    return DB_SUCCESS;
+  }
+
+  uint32_t page_no = fseg_get_nth_frag_page_no(inode, n);
+  if (header && page_no == page_get_page_no(page_align(header)))
+     return DB_SUCCESS;
+
+  err= fseg_free_page_low(inode, iblock, space, page_no, mtr
+#ifdef BTR_CUR_HASH_ADAPT
+                          , ahi
+#endif /* BTR_CUR_HASH_ADAPT */
+                          );
+  if (err != DB_SUCCESS)
+    return err;
+  buf_page_free(space, page_no, mtr);
+
+  if (free_inode)
+  {
+    n= fseg_find_last_used_frag_page_slot(inode);
+    if (UNIV_UNLIKELY(n == ULINT_UNDEFINED))
+    {
+      fsp_free_seg_inode(space, inode, iblock, mtr);
+      return DB_SUCCESS;
+    }
+  }
```

Comment:
Sorry, I realized that `fsp_free_seg_inode()` requires that the allocation bitmap be all clear upfront. We have to retain the current order because of that.